### PR TITLE
[gbc c++] Use init. list for populating enum maps

### PR DESCRIPTION
**Pull request by ara-ayvazyan**
Currently, `boost::assign::map_list_of` is used to initialize C++ enum maps. This change makes use of [list initialization](http://en.cppreference.com/w/cpp/language/list_initialization) available in `c++11` which is more appropriate and avoids extra dependency on template-heavy boost headers.

### CODE REVIEW

File: compiler/tests/generated/alloc_ctors/attributes_types.cpp [hunk 1]
```diff
@@ -10,9 +10,10 @@ namespace tests
     namespace Enum
     {
         const
-        std::map<std::string, enum Enum> _name_to_value_Enum =
-            boost::assign::map_list_of<std::string, enum Enum>
-                ("Value1", Value1);
+        std::map<std::string, enum Enum> _name_to_value_Enum
+            {
+                { "Value1", Value1 }
```
Comment (chwarr) [inline compiler/tests/generated/alloc_ctors/attributes_types.cpp]: Inserting into a map is going to be a bit faster if the keys are already in order. Can we sort them in Haskell?

File: compiler/tests/generated/alloc_ctors/attributes_types.cpp [hunk 2]
```diff
@@ -10,9 +10,10 @@ namespace tests
...
+            };
 
         const
         std::map<enum Enum, std::string> _value_to_name_Enum =
```
Comment (chwarr) [inline compiler/tests/generated/alloc_ctors/attributes_types.cpp]: Can we generated the reverse map in Haskell as well?

**Comment by ara-ayvazyan** [comment]
The test failure is in unrelated grpc test.

File: compiler/src/Language/Bond/Syntax/Util.hs [hunk 3]
```diff
@@ -231,3 +232,12 @@ resolveAlias Alias {..} args = fmapType resolveParam aliasType
     paramsArgs = zip declParams args
 resolveAlias _ _ = error "resolveAlias: impossible happened."
 
+
+
+-- | Fill in values for constants w/o explicitly specified value
+constNameValues :: [Constant] -> [(String, Int)]
+constNameValues constants = nameValues 0 constants
```
Comment (chwarr) [inline compiler/src/Language/Bond/Syntax/Util.hs]: `reifyEnumValues` ?

File: compiler/src/Language/Bond/Codegen/Cpp/Types_cpp.hs [hunk 4]
```diff
@@ -89,9 +91,13 @@ types_cpp cpp file _imports declarations = ("_types.cpp", [lt|
...
             return true;
         }
+
     } // namespace #{declName}
     } // namespace _bond_enumerators|]
       where
-        constant Constant {..} = [lt|{ "#{constantName}", #{constantName} }|]
+        nameValueConst Constant {..} = [lt|{ "#{constantName}", #{constantName} }|]
+        valueNameConst (name, _) = [lt|{ #{name}, "#{name}" }|]
+        enumConstByName = sortBy (\x y -> constantName x `compare` constantName y) enumConstants
```
Comment (chwarr) [inline compiler/src/Language/Bond/Codegen/Cpp/Types_cpp.hs]: Does `sortBy (comparing constantName) enumConstants` or `sortOn constantName enumConstants` work?

File: compiler/src/Language/Bond/Codegen/Cpp/Types_cpp.hs [hunk 5]
```diff
@@ -89,9 +91,13 @@ types_cpp cpp file _imports declarations = ("_types.cpp", [lt|
...
+        enumConstByValue = sortBy (\x y -> snd x `compare` snd y) $ constNameValues enumConstants
```
Comment (chwarr) [inline compiler/src/Language/Bond/Codegen/Cpp/Types_cpp.hs]: Similarly, does `sortBy (comparing snd) $ constNameValues enumConstants` or `sortOn snd $ constNameValues enumConstants` work?

File: compiler/src/Language/Bond/Syntax/Util.hs (hunk 3 above)
Comment (ara-ayvazyan) [reply compiler/src/Language/Bond/Syntax/Util.hs]: Updated.

**Pull request merged**